internal/extender/resource.go

```diff
@@ -336,92 +337,66 @@ func (s *SparkSchedulerExtender) potentialNodes(availableNodesSchedulingMetadata
+				return "", failureInternal, werror.WrapWithContextParams(ctx, err, "failed to reserve node for executor")
 			}
-			// We might have created a demand object for this executor when we were under min count, so we should remove if it exists
 			s.removeDemandIfExists(ctx, executor)
-			return node, successScheduledExtraExecutor, nil
+			return resultNode, success, nil
 		}
-		return "", outcome, unboundResErr
-	}
-	// the reservation to be selected for the current executor needs to be on a node that the extender has received from kube-scheduler
-	nodeToReservation := make(map[string]string, len(unboundReservations))
-	for _, reservationName := range unboundReservations {
-		nodeToReservation[resourceReservation.Spec.Reservations[reservationName].Node] = reservationName
 	}
-	var unboundReservation string
-	for _, name := range nodeNames {
-		if reservation, ok := nodeToReservation[name]; ok {
-			unboundReservation = reservation
-			break
-		}
+
+	// Else, check if you still can have an executor, and if yes, reschedule
+	freeExecutorSpots, err := s.resourceReservationManager.GetFreeExecutorSpots(ctx, executor)
+	if err != nil {
+		return "", failureInternal, werror.WrapWithContextParams(ctx, err, "error when checking for free executor spots")
 	}
-	copyResourceReservation := resourceReservation.DeepCopy()
-
-	if unboundReservation == "" {
-		// no nodes for the unbound reservations exists in nodeNames, this might be because the node for the reservations are terminated
-		// try to reschedule the executor, breaking FIFO, but preventing executor starvation
-		// we are guaranteed len(unboundResourceReservations) > 0
-		unboundReservation = unboundReservations[0]
-		driver, err := s.podLister.getDriverPod(ctx, executor)
-		if err != nil {
-			return "", failureInternal, err
-		}
-		sparkResources, err := sparkResources(ctx, driver)
+	if freeExecutorSpots > 0 {
+		isExtraExecutor := len(unboundReservationNodes) > 0 // Dynamic allocation executor since there were no free unbound reservations
+		nodeName, outcome, err := s.rescheduleExecutor(ctx, executor, nodeNames, isExtraExecutor)
 		if err != nil {
-			return "", failureInternal, err
+			return "", outcome, werror.WrapWithContextParams(ctx, err, "failed to reschedule executor")
 		}
-		node, outcome, err := s.rescheduleExecutor(ctx, executor, nodeNames, sparkResources, true)
+		err = s.resourceReservationManager.ReserveForExecutor(ctx, executor, nodeName)
 		if err != nil {
-			return "", outcome, err
+			return "", failureInternal, werror.WrapWithContextParams(ctx, err, "failed to reserve node for rescheduled executor")
 		}
-		svc1log.FromContext(ctx).Info("rescheduling executor", svc1log.SafeParam("reservationName", unboundReservation), svc1log.SafeParam("node", node))
-		reservation := copyResourceReservation.Spec.Reservations[unboundReservation]
-		reservation.Node = node
-		copyResourceReservation.Spec.Reservations[unboundReservation] = reservation
+		s.removeDemandIfExists(ctx, executor)
```

Comment:
It seems surprising that this is buried deep here - feels like it should just be an invariant very high up in the code, that if we are about to return a successful schedule, we `removeDemandIfExists`